### PR TITLE
More thorough CoalesceUnaryMappable

### PR DESCRIPTION
**Pull request by wemrysi**
Improves `CoalesceUnaryMappable` to coalesce source `Map`s into an `AutoJoin{2,3}`. This ended up requiring better provenance normalization for performance of computing provenance on large `MapFunc`s.

Also updated `InlineNullary` to be more performant as it didn't require the amount of traversing it was performing.

### CODE REVIEW

**Comment by wemrysi** [comment]
[ch2085]

**Comment by djspiewak** [comment]
Version label. :-)

**Comment by alissapajer** [comment]
Confirming we don't need this against backport?

**Comment by alissapajer** [comment]
@wemrysi failed to build due to an unused import.

**Comment by djspiewak** [comment]
> Confirming we don't need this against backport?

Confirmed. I mean, it would have been nice, but this work was already done against master *and* 1.0 is effectively frozen, so we're good.

**Comment by alissapajer** [comment]
It's cool, it'll land in 1.1!

**Comment by djspiewak** [comment]
> It's cool, it'll land in 1.1!

Probably more like 1.0.1 :-)

File: qscript/src/main/scala/quasar/qscript/provenance/Prov.scala [hunk 1]
```diff
@@ -95,8 +95,39 @@ trait Prov[D, I, P] {
     implicit def provConjunctionSemiLattice(implicit D: Equal[D], I: Equal[I]): SemiLattice[P @@ Conjunction] =
       new SemiLattice[P @@ Conjunction] {
         def append(a: P @@ Conjunction, b: => P @@ Conjunction) =
-          Conjunction(distinctConjunctions(both(a.unwrap, b.unwrap)))
+          Conjunction(zipEq(a.unwrap, b.unwrap))
+
+        def zipEq(l: P, r: P): P = {
```
Comment (alissapajer) [inline qscript/src/main/scala/quasar/qscript/provenance/Prov.scala]: Could you add a comment about what this function does? I think it's returning the longest equal prefix `≺:`'d together, followed by the tails conjunctioned together?

**Comment by alissapajer** [comment]
@wemrysi real qsu test failure.

File: qscript/src/test/scala/quasar/qscript/provenance/ProvSpec.scala [hunk 2]
```diff
@@ -135,6 +135,32 @@ final class ProvSpec extends Qspec with ProvFGenerator {
     }
   }
 
+  "conjunction" >> {
```
Comment (alissapajer) [inline qscript/src/test/scala/quasar/qscript/provenance/ProvSpec.scala]: Neat tests!

File: qsu/src/test/scala/quasar/qsu/CoalesceUnaryMappableSpec.scala [hunk 3]
```diff
@@ -75,5 +76,60 @@ object CoalesceUnaryMappableSpec extends Qspec with QSUTTypes[Fix] with TreeMatc
...
+            qsu.map(
+              qsu.read(dataA),
+              rec.ProjectKeyS(rec.Hole, "X")),
+            rec.MakeMapS("A", rec.Hole)),
+          qsu.map(
+            qsu.map(
+              qsu.read(dataB),
+              rec.ProjectKeyS(rec.Hole, "Y")),
+            rec.MakeMapS("B", rec.Hole)),
+          mf.ConcatMaps(mf.LeftSide, mf.RightSide)))
```
Comment (alissapajer) [inline qsu/src/test/scala/quasar/qsu/CoalesceUnaryMappableSpec.scala]: In the original, we are joining on projections of the data, while in the expected results we are joining on the data itself and then projecting _after_ the join. If this is a valid transformation, then both these graphs have the same provenance. Do they have the same provenance in practice?

File: qscript/src/main/scala/quasar/qscript/provenance/Prov.scala (hunk 1 above)
Comment (wemrysi) [reply qscript/src/main/scala/quasar/qscript/provenance/Prov.scala]: Aye, I'll add a comment and rename to `zipWhileEqual` as that is a bit more descriptive.

File: qsu/src/test/scala/quasar/qsu/CoalesceUnaryMappableSpec.scala (hunk 3 above)
Comment (wemrysi) [reply qsu/src/test/scala/quasar/qsu/CoalesceUnaryMappableSpec.scala]: With the changes to provenance computation in this PR, they do.

**Comment by alissapajer** [comment]
@wemrysi an integration test timed out in appveyor
```
[error]   x [qa_s08] groupBy with avg, max, min, count, sum, and arbitrary [./reduce/groupBy_with_avg_max_min_count_sum_and_arbitrary.test] (3 minutes 25 seconds, 222 ms)
[error]    Timed out. (Sql2QueryRegressionSpec.scala:156)
[info] Sql2QueryRegressionSpec
[error] Failed: Total 915, Failed 1, Errors 0, Passed 889, Skipped 25, Pending 17
[error] Failed tests:
[error] 	quasar.regression.Sql2QueryRegressionSpec
[error] (it / Test / testOnly) sbt.TestsFailedException: Tests unsuccessful
```
haven't seen that time out before.

File: qsu/src/test/scala/quasar/qsu/CoalesceUnaryMappableSpec.scala (hunk 3 above)
Comment (alissapajer) [reply qsu/src/test/scala/quasar/qsu/CoalesceUnaryMappableSpec.scala]: Do the other test expectations you updated cover this case, or is there another test that can be added to test this explicitly?

**Comment by wemrysi** [comment]
@alissapajer I'm going to restart appveyor, that test that timed out is actually faster than master on travis, so hoping it was an environmental issue on appveyor.

File: qsu/src/test/scala/quasar/qsu/CoalesceUnaryMappableSpec.scala (hunk 3 above)
Comment (wemrysi) [reply qsu/src/test/scala/quasar/qsu/CoalesceUnaryMappableSpec.scala]: I don't think so, I like the way you described the difference, so I'll add a test that exercises that equality.

**Comment by wemrysi** [comment]
@alissapajer Ready for another look!

**Comment by alissapajer** [comment]
@wemrysi scalacheck qscript test failed in appveyor. Is it related to this work, or should we open a ticket and restart the build?
```
[error]   x scalaz.semilattice.band.semigroup.associative (172 ms)
[error]    Falsified after 43 passed tests.
[error]    > ARG_0: Fix(Then(Fix(Value(-1)),Fix(Value(2147483647))))
[error]    > ARG_1: Fix(Value(2147483647))
[error]    > ARG_2: Fix(Value(0))
[error]    The seed is WFyR8UTkqMhEr2WMVy-N-RYmZxSf84VvDnC_0Pv0WBI=
[error] Failed: Total 203, Failed 1, Errors 0, Passed 202
[error] Failed tests:
[error] 	quasar.qscript.provenance.ProvSpec
```

**Comment by wemrysi** [comment]
@alissapajer Definitely isn't related to this PR and is also disconcerting:

```
scala> val p = Prov[Char, Int, Fix[ProvF[Char, Int, ?]]]
p: quasar.qscript.provenance.Prov[Char,Int,matryoshka.data.Fix[[γ$0$]quasar.qscript.provenance.ProvF[Char,Int,γ$0$]]] = quasar.qscript.provenance.Prov$$anon$2@12780e47

scala> val a0 = p.thenn(p.value(-1), p.value(Int.MaxValue))
a0: matryoshka.data.Fix[[γ$0$]quasar.qscript.provenance.ProvF[Char,Int,γ$0$]] = Fix(Then(Fix(Value(-1)),Fix(Value(2147483647))))                                                                                                              

scala> val a1 = p.value(Int.MaxValue)
a1: matryoshka.data.Fix[[γ$0$]quasar.qscript.provenance.ProvF[Char,Int,γ$0$]] = Fix(Value(2147483647))

scala> val a2 = p.value(0)
a2: matryoshka.data.Fix[[γ$0$]quasar.qscript.provenance.ProvF[Char,Int,γ$0$]] = Fix(Value(0))

scala> import scalaz.Scalaz._
import scalaz.Scalaz._

scala> import p.implicits._
import p.implicits._

scala> (a0 |+| a1) |+| a2
res2: matryoshka.data.Fix[[γ$0$]quasar.qscript.provenance.ProvF[Char,Int,γ$0$]] = Fix(Then(Fix(Then(Fix(Then(Fix(Value(-1)),Fix(Value(2147483647)))),Fix(Value(2147483647)))),Fix(Value(0))))                                                 

scala> a0 |+| (a1 |+| a2)
res3: matryoshka.data.Fix[[γ$0$]quasar.qscript.provenance.ProvF[Char,Int,γ$0$]] = Fix(Then(Fix(Then(Fix(Value(-1)),Fix(Value(2147483647)))),Fix(Then(Fix(Value(2147483647)),Fix(Value(0))))))                                                 

scala> res2 === res3
res4: Boolean = true

scala> res3 === res2
res5: Boolean = true

scala> (a0 |+| (a1 |+| a2)) === ((a0 |+| a1) |+| a2)
res6: Boolean = true

**Comment by wemrysi** [comment]
@alissapajer ticket created and appveyor restarted.

**Pull request merged**